consumer: cleanup RDY handling; fix pausing

mreiferson opened the pull request:
this addresses the issue reported in #222

mreiferson commented:
RFR

ploxiln commented on mock_test.go:
```diff
@@ -111,25 +112,25 @@ func (n *mockNSQD) handle(conn net.Conn) {
 				_, err := io.ReadFull(rdr, l)
 				if err != nil {
-					log.Printf(err.Error())
+					n.t.Logf(err.Error())
```
this should use `Log()` instead of `Logf()` (and 2 other cases below)

mreiferson replied:
OK, fine

ploxiln commented:
I noticed that `rdyRetryMtx` and `backoffMtx` can be plain Mutex instead of RWMutex

ploxiln commented:
This is honestly the part of NSQ that I'm least familiar with ... the technical details of how ready-count is "refilled" and such ... I'll keep looking and slowly digesting though :)

mreiferson commented:
Addressed comments.

> the technical details of how ready-count is "refilled" and such

Yea, this is just old and outdated code that's already been [addressed in `pynsq`](https://github.com/nsqio/pynsq/pull/179). See https://github.com/nsqio/nsq/pull/404 for relevant context.

mreiferson commented on mock_test.go:
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
+
+	timeoutCh := time.After(500 * time.Millisecond)
+	pauseCh := time.After(50 * time.Millisecond)
+	unpauseCh := time.After(75 * time.Millisecond)
```
This is a little fragile and time dependent, but should be reasonably stable.

ploxiln commented on conn.go:
```diff
@@ -207,13 +206,12 @@ func (c *Conn) RDY() int64 {
 
 // LastRDY returns the previously set RDY count
```
I guess this function is somewhat deprecated now?

mreiferson replied:
Yea

ploxiln commented on mock_test.go:
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
 }
+
+func TestConsumerPause(t *testing.T) {
+	msgIDGood := MessageID{'1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'a', 's', 'd', 'f', 'g', 'h'}
```
not an expected message id, as it has characters that are not hex ... not sure if it matters at all

very minor idea, alternate way to write this:

```go
	var msgId MessageID
	copy(msgId[:], []byte("0123456789abcdef"))
```

mreiferson replied:
copied from previous tests and yes, it doesn't matter

ploxiln commented on mock_test.go:
```diff
@@ -470,3 +469,73 @@ func TestConsumerBackoffDisconnect(t *testing.T) {
+	script := []instruction{
+		// SUB
+		instruction{0, FrameTypeResponse, []byte("OK")},
+		// IDENTIFY
```
I think IDENTIFY is before SUB?

mreiferson replied:
Fixing

ploxiln commented:
looks good 👍
